providers/configcat/src/main/java/dev/openfeature/contrib/providers/configcat/ConfigCatProvider.java

```diff
@@ -0,0 +1,205 @@
+package dev.openfeature.contrib.providers.configcat;
+
+import com.configcat.ConfigCatClient;
+import com.configcat.EvaluationDetails;
+import com.configcat.User;
+import dev.openfeature.sdk.EvaluationContext;
+import dev.openfeature.sdk.EventProvider;
+import dev.openfeature.sdk.Metadata;
+import dev.openfeature.sdk.ProviderEvaluation;
+import dev.openfeature.sdk.ProviderEventDetails;
+import dev.openfeature.sdk.ProviderState;
+import dev.openfeature.sdk.Value;
+import dev.openfeature.sdk.exceptions.GeneralError;
+import dev.openfeature.sdk.exceptions.ProviderNotReadyError;
+import lombok.Getter;
+import lombok.SneakyThrows;
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.ArrayList;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * Provider implementation for ConfigCat.
+ */
+@Slf4j
+public class ConfigCatProvider extends EventProvider {
+
+    @Getter
+    private static final String NAME = "ConfigCat";
+
+    public static final String PROVIDER_NOT_YET_INITIALIZED = "provider not yet initialized";
+    public static final String UNKNOWN_ERROR = "unknown error";
+
+    private ConfigCatProviderConfig configCatProviderConfig;
+
+    @Getter
+    private ConfigCatClient configCatClient;
+
+    @Getter
+    private ProviderState state = ProviderState.NOT_READY;
+
+    private AtomicBoolean isInitialized = new AtomicBoolean(false);
+
+    /**
+     * Constructor.
+     * @param configCatProviderConfig configCatProvider Config
+     */
+    public ConfigCatProvider(ConfigCatProviderConfig configCatProviderConfig) {
+        this.configCatProviderConfig = configCatProviderConfig;
+    }
+
+    /**
+     * Initialize the provider.
+     * @param evaluationContext evaluation context
+     * @throws Exception on error
+     */
+    @Override
+    public void initialize(EvaluationContext evaluationContext) throws Exception {
+        boolean initialized = isInitialized.getAndSet(true);
+        if (initialized) {
+            throw new GeneralError("already initialized");
+        }
+        super.initialize(evaluationContext);
+        configCatClient = ConfigCatClient.get(configCatProviderConfig.getSdkKey(),
+            configCatProviderConfig.getOptions());
+        configCatProviderConfig.postInit();
+        state = ProviderState.READY;
+        log.info("finished initializing provider, state: {}", state);
+
+        configCatClient.getHooks().addOnClientReady(() -> {
```

Comment:
I'm not 100% sure that if you add the addOnClientReady hook after the initialization, it will fire in all cases. 
I'd move the hook configuration inside the ConfigCatClient.get call, like here: https://configcat.com/docs/sdk-reference/java/#hooks

After that, I think the state = ProviderState.READY; can be removed after the configCatProviderConfig.postInit(); call. I think, the clientReady should set the state only.

Reply:
Ok, removed the _ClientReady_ event, as not seems it will be correct and necessary in all cases.
The non-ready state is at least temporary until initialization for corner cases (other comment).